core/src/main/java/feast/core/model/FeatureTable.java

```diff
@@ -0,0 +1,323 @@
+  @JoinColumn(name = "batch_source_id", nullable = true)
+  private FeatureSource batchSource;
+
+  // Autoincrementing version no. of this Feature Table.
+  // Autoincrements every update made to the feature set.
+  @Column(name = "revision", nullable = false)
+  private int revision;
+
+  public FeatureTable() {
+    this.revision = 1;
+  };
+
+  /**
+   * Construct Feature from Protobuf spec representation in the given project with entities
+   * registered in entity repository.
+   *
+   * @param projectName the name of the project that the constructed Feature table belongs.
+   * @param spec the Protobuf spec to contruct the Feature from.
+   * @param entityRepo {@link EntityRepository} used to resolve entity names.
+   * @throws IllegalArgumentException if the Protobuf spec provided is invalid.
+   * @return constructed Feature from the given Protobuf spec.
+   */
+  public static FeatureTable fromProto(
+      String projectName, FeatureTableSpec spec, EntityRepository entityRepo) {
+    FeatureTable table = new FeatureTable();
+    table.setName(spec.getName());
+    table.setProject(new Project(projectName));
+
+    Set<FeatureV2> features =
+        spec.getFeaturesList().stream()
+            .map(featureSpec -> FeatureV2.fromProto(table, featureSpec))
+            .collect(Collectors.toSet());
+    table.setFeatures(features);
+
+    Set<EntityV2> entities =
+        FeatureTable.resolveEntities(
+            projectName, spec.getName(), entityRepo, spec.getEntitiesList());
+    table.setEntities(entities);
+
+    String labelsJSON = TypeConversion.convertMapToJsonString(spec.getLabelsMap());
+    table.setLabelsJSON(labelsJSON);
+
+    table.setMaxAgeSecs(spec.getMaxAge().getSeconds());
+
+    // Configure sources only if set
+    if (!spec.getStreamSource().equals(FeatureSourceSpec.getDefaultInstance())) {
+      table.setStreamSource(FeatureSource.fromProto(spec.getStreamSource()));
+    }
+    if (!spec.getBatchSource().equals(FeatureSourceSpec.getDefaultInstance())) {
+      table.setBatchSource(FeatureSource.fromProto(spec.getBatchSource()));
+    }
+
+    return table;
+  }
+
+  /**
+   * Update the FeatureTable from the given Protobuf representation.
+   *
+   * @param spec the Protobuf spec to update the FeatureTable from.
+   * @throws IllegalArgumentException if the update will make prohibited changes.
+   */
+  public void updateFromProto(FeatureTableSpec spec) {
+    // Check for prohibited changes made in spec:
+    // - Name cannot be changed
+    if (!getName().equals(spec.getName())) {
+      throw new IllegalArgumentException(
+          String.format(
+              "Updating the name of a registered FeatureTable is not allowed: %s to %s",
+              getName(), spec.getName()));
+    }
+    // - Entities cannot be changed
+    List<String> entityNames =
+        getEntities().stream().map(EntityV2::getName).collect(Collectors.toList());
+    if (!entityNames.containsAll(spec.getEntitiesList())) {
```

Comment:
does it check for deletion of entity?

Reply:
By deletion, you mean from [entity1, entity2] to [entity1]?

Reply:
yes

Reply:
Good catch, it should have been `containsAll` both ways. Updated this check. Added a test [here](https://github.com/feast-dev/feast/pull/1019/files#diff-5dcec7c4d3f743c906004faffa9b5018R1121) as well.